L1Trigger/TrackFindingTracklet/plugins/ProducerKFin.cc

```diff
@@ -0,0 +1,261 @@
+        vector<StubKFin> stubs;
+        const int nStubs = accumulate(ttTrackRefs.begin(), ttTrackRefs.end(), 0, [](int& sum, const TTTrackRef& ttTrackRef){ return sum += ttTrackRef->getStubRefs().size(); });
+        stubs.reserve(nStubs);
+        int trackId(0);
+        vector<int> numLayerStubs(setup_->numLayers(), 0);
+        for (const TTTrackRef& ttTrackRef : ttTrackRefs) {
+          // cut on more then 256 tracks
+          if (trackId >= dataFormats_->format(Variable::trackId, Process::kfin).range())
+            continue;
+          // get rz parameter
+          double cot = ttTrackRef->tanL();
+          double zT = ttTrackRef->z0() + setup_->chosenRofZ() * cot;
+          int binEta(-1);
+          for (; binEta < setup_->numSectorsEta(); binEta++)
+            if (zT < sinh(setup_->boundarieEta(binEta + 1)) * setup_->chosenRofZ())
+              break;
+          // cut on outer eta sector boundaries
+          if (binEta == -1 || binEta == setup_->numSectorsEta())
+            continue;
+          cot -= setup_->sectorCot(binEta);
+          zT -= setup_->sectorCot(binEta) * setup_->chosenRofZ();
+          // cut on eta and |z0| < 15 cm
+          if (!dfZT.inRange(zT) || !dfCot.inRange(cot))
+            continue;
+          const int binZT = dfZT.toUnsigned(dfZT.integer(zT));
+          const int binCot = dfCot.toUnsigned(dfCot.integer(cot));
+          const vector<int>& layerEncoding = layerEncoding_->layerEncoding(binEta, binZT, binCot);
+          // get rphi parameter
+          double qOverPt = ttTrackRef->rInv() / 2.;
+          // calculcate track phi at radius hybridChosenRofPhi with respect to phi sector centre
+          double phiT = deltaPhi(ttTrackRef->phi() - setup_->hybridChosenRofPhi() * qOverPt - ttTrackRef->phiSector() * setup_->baseRegion());
+          const int sectorPhi = phiT < 0. ? 0 : 1; // dirty hack
+          phiT -= (sectorPhi - .5) * setup_->baseSector();
+          // cut on nonant size and pt
+          if (!dfPhiT.inRange(phiT) || !dfQoverPt.inRange(qOverPt))
+            continue;
+          // loop over stubs
+          TTBV hitPattern(0, setup_->numLayers());
+          vector<int> layerMap(setup_->numLayers(), 0);
+          for (const TTStubRef& ttStubRef : ttTrackRef->getStubRefs()) {
+            const GlobalPoint& gp = setup_->stubPos(ttStubRef);
+            const double r = gp.perp() - setup_->hybridChosenRofPhi();
+            const double phi = deltaPhi(gp.phi() - (ttTrackRef->phi() - qOverPt * gp.perp()));
+            const double z = gp.z() - (ttTrackRef->z0() + ttTrackRef->tanL() * gp.perp());
+            double layer = distance(layerEncoding.begin(), find(layerEncoding.begin(), layerEncoding.end(), setup_->layerId(ttStubRef)));
+            if (layer >= setup_->numLayers())
+              layer = setup_->numLayers() - 1;
+            // cut on phi and z residuals
+            if (!dfPhi.inRange(phi) || !dfZ.inRange(z))
+              continue;
```

Comment:
Are these residual cuts ever tighter than those applied in the Tracklet pattern reco (specified in TrackFindingTracklet/interface/Settings.h starting with at line with variable "rphiMatchCut_"). It would be silly to use loose cuts in Tracklet, and then kill the looser stubs in KF.